pkg/endpointmanager/host.go

```diff
@@ -58,21 +47,31 @@ func (mgr *endpointManager) OnAddNode(newNode *slim_corev1.Node,
 func (mgr *endpointManager) OnUpdateNode(oldNode, newNode *slim_corev1.Node,
 	swg *lock.StoppableWaitGroup) error {
 
-	oldNodeLabels := oldNode.GetLabels()
+	var oldNodeLabels map[string]string
+	// Endpoint's label update logic rejects a request if any of the old labels are
+	// not present in the endpoint manager's state. Overwrite the old labels to nil
+	// for the first node update to avoid node labels being outdated indefinitely (GH-29649).
+	if oldNode == nil || isFirstNodeUpdate {
+		oldNodeLabels = make(map[string]string)
+	} else {
+		oldNodeLabels = oldNode.GetLabels()
+	}
 	newNodeLabels := newNode.GetLabels()
+	// Set the labels early so host endpoint is created with latest labels.
+	node.SetLabels(newNodeLabels)
 
 	nodeEP := mgr.GetHostEndpoint()
 	if nodeEP == nil {
-		log.Error("Host endpoint not found")
+		log.Debug("Host endpoint not found")
 		return nil
 	}
 
-	node.SetLabels(newNodeLabels)
-
-	err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
-	if err != nil {
+	if err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s); err != nil {
+		log.WithError(err).Error("Unable to update host endpoint labels")
```

Comment:
Assuming this error message doesn't show up in regular lifecycle churn, I'm OK with this being an error. That said, the lifecycles of nodes and endpoints are theoretically independent -- I'm not sure what the invariants are here.

Reply:
Yeah, this does not show up in regular lifecycle churn. This error message indicate a real failure if the node labels are not updated and host firewall could not applied correctly. The downstream does not log this error so it was difficult to debug this issue in our env.